### DISCUSSION TITLE
How are requests queued?

### DISCUSSION THREAD

**Discussion by raqibhayder** [fastapi/fastapi#4188] [Questions] [answered]
I am trying to understand how FastAPI/Gunicorn handles and queues requests.

**Client contract**: Once the client dispatches a request, it needs to receive a response in 5 seconds. If the client does not receive a response within 5 seconds, the request is no longer useful and the client stops listening for the response.

**Scenario:**
Consider the following scenario:
1. Let's say we have 4 workers (using Gunicorn as process manager with UvicornWorkers; using [tiangolo's docker image](https://github.com/tiangolo/uvicorn-gunicorn-fastapi-docker))
2. Client sends 100 concurrent requests
3. Each request takes 5 seconds to complete
4. We process the first 4 requests in 5 seconds
5. But by that time, the remaining 96 requests accepted in the queue (I am not sure if queueing happens or how it works, would appreciate it if you can explain this) have already spent 5 seconds there and breached the client contract. Processing these 96 requests is a waste of resources as the client is no longer listening for them. 

**Question:**
1. How would I discard these 96 requests i.e. if each request has been waiting for more than 5 seconds in the queue, drop them?

**Comment by jgould22** [comment, marked as answer]
FastAPI is an asynchronous framework, it will not queue requests in Gunicorn as if you were using sync workers which handle a single request at a time.

Each uvicorn worker can support hundreds or thousands of requests per worker if they are not doing any intensive processing (ie work that might block the event loop).  

If each request is process intensive then you should not be doing that processing inside the FastAPI framework, pass it off to something like Celery.

**Reply by padrepitufo** [reply]
This doesn't necessarily answer the question for all scenarios.  Yes, there are cases/workloads where moving over to a message queue makes sense.  However, there exist many situations appropriate for an API where the service serving the request may fall into a degraded state (e.g due to the amount of requests coming in), and it could have degraded so much so that continuing to serve the request or continuing to take more requests may result in an undesirable outcome for the client and/or server.

It seems like the question presented is part of capacity planning,  graceful degradation, etc.  That's where I am at least and understanding a bit more about FastAPI would help.

**Reply by jgould22** [reply]
Fair enough.

The question was 
>  How would I discard these 96 requests i.e. if each request has been waiting for more than 5 seconds in the queue, drop them?

There is [no real queue in gunicorn](https://docs.gunicorn.org/en/stable/design.html) in this instance, it just spawns python application processes that all listen on a socket. When a request comes in it is up to the OS to wake a process and accept the connection. 

Uvicorn itself has [some settings ](https://www.uvicorn.org/settings/#resource-limits)to control backlog and concurrency however they [are not available when using gunicorn ](https://www.uvicorn.org/deployment/#gunicorn)

If you block the uvicorn workers or they are overloaded they will not be able to respond to the request. In that instance it becomes a waiting game to see who times out first, gunicorn which may kill the worker because it is not responding, or the client who is tired of waiting.

So the correct solution seems to be handle your own queue and decouple it from your web framework so the web framework is always able to respond and gracefully (ie respond with a 503 or 429 or whatever).  The number of simultaneous requests a single worker can handle will be application dependent and something you would have to benchmark.
